### PR TITLE
Implements a generic LinkProcessingExtractor

### PR DESCRIPTION
**Pull request by leogrim**
@ymatsuda @stkem @yingjieMiao 

This extractor builds upon DefaultExtractor to process the links found within a scraped page, both for side effects (e.g. report a canonical url) or to aggregate their content with the main page's (e.g. fetch "About" page). It becomes the new default extractor.

Right now, the only links being processed are the "About" links (aggregate with main content) but I have tried to build it so that other heuristics (e.g. report canonical urls or fetch menu for a restaurant page) can easily be added to LinkProcessingExtractorProvider

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/scraper/extractor/LinkProcessingExtractor.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        case _ => None
+      }
+    }
+  }
+
+  private def isAbout(baseUrl: String, aboutUrl: String): Boolean = {
+    val aboutPages = Seq("about", "aboutus", "about-us", "about_us", "apropos", "a-propos", "a_propos")
+    val aboutSuffix = ("/(" + aboutPages.mkString("|") + """)(\.[a-z]{2,4})?/?$""").r
+    aboutUrl.startsWith(baseUrl) && aboutSuffix.pattern.matcher(aboutUrl).find
+  }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/scraper/extractor/LinkProcessingExtractor.scala:67]: Calling this a "heuristic" makes it sound much smarter than it is.

File: kifi-backend/modules/shoebox/app/com/keepit/scraper/extractor/LinkProcessingExtractor.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    if (keywords.isEmpty) None else Some(keywords)
+  }
+}
+
+@Singleton
+class LinkProcessingExtractorProvider @Inject() (httpFetcher: HttpFetcher, db: Database, urlPatternRuleRepo: UrlPatternRuleRepo) extends ExtractorProvider {
+  def isDefinedAt(uri: URI) = true
+  def apply(uri: URI) = new LinkProcessingExtractor(uri.toString, Scraper.maxContentChars, DefaultExtractorProvider.htmlMapper, processLink(uri), httpFetcher, db, urlPatternRuleRepo)
+
+  private def processLink(uri: URI): Link => Option[String] = {
```
Comment (ymatsuda) [inline kifi-backend/modules/shoebox/app/com/keepit/scraper/extractor/LinkProcessingExtractor.scala]: I think you can use currying to make this a little nicer.

**Comment by ymatsuda** [comment]
lgtm. awesome!

File: kifi-backend/modules/shoebox/app/com/keepit/scraper/extractor/LinkProcessingExtractor.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/scraper/extractor/LinkProcessingExtractor.scala]: I'm curious, do you know if partial application means partial evaluation as well?

**Pull request merged**